drivers/input/input_esp32_touch_sensor.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2023 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_touch_sensor
+
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <esp_err.h>
+#include <soc/soc_pins.h>
+#include <soc/periph_defs.h>
+#include <hal/touch_sensor_types.h>
+#include <hal/touch_sensor_hal.h>
+#include <driver/rtc_io.h>
+#include <esp_intr_alloc.h>
+
+LOG_MODULE_REGISTER(espressif_touch_sensor, CONFIG_INPUT_LOG_LEVEL);
+
+#define SCAN_DONE_MAX_COUNT (5)
+
+struct touch_sensor_channel_config {
+	int32_t channel_num;
+	int32_t channel_sens;
+	uint32_t zephyr_code;
+};
+
+struct touch_sensor_config {
+	uint32_t debounce_interval_ms;
+	const int num_channels;
+	const int volt_refh;
+	const int volt_refl;
+	const int volt_atten;
+	const int filter_mode;
+	const int filter_debounce_cnt;
+	const int filter_noise_thr;
+	const int filter_jitter_step;
+	const int filter_smooth_level;
+	const struct touch_sensor_channel_config *channel_cfg;
+};
+
+struct touch_sensor_channel_data {
+	const struct device *dev;
+	struct k_work_delayable work;
+	uint32_t status;
+#if defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3)
+	uint32_t last_status;
+#endif /* defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3) */
+};
+
+struct touch_sensor_data {
+	uint32_t rtc_intr_msk;
+	struct touch_sensor_channel_data *channel_data;
+};
+
+static void touch_sensor_interrupt_cb(void *arg)
+{
+	const struct device *dev = (struct device *)arg;
+	struct touch_sensor_data *dev_data = dev->data;
+	const struct touch_sensor_config *dev_cfg = dev->config;
+	const struct touch_sensor_channel_config *channel_cfg;
+	const int num_channels = dev_cfg->num_channels;
+	uint32_t pad_status;
+
+#if defined(CONFIG_SOC_SERIES_ESP32)
+	touch_hal_intr_clear();
+
+#elif defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3)
+	static uint8_t scan_done_counter;
+
+	touch_pad_intr_mask_t intr_mask = touch_hal_read_intr_status_mask();
+
+	if (intr_mask & TOUCH_PAD_INTR_MASK_SCAN_DONE) {
+		if (scan_done_counter++ == SCAN_DONE_MAX_COUNT) {
+			touch_hal_intr_disable(TOUCH_PAD_INTR_MASK_SCAN_DONE);
+			for (int i = 0; i < num_channels; i++) {
+				channel_cfg = &dev_cfg->channel_cfg[i];
+
+				/* Set interrupt threshold */
+				uint32_t benchmark_value;
+
+				touch_hal_read_benchmark(channel_cfg->channel_num,
+					&benchmark_value);
+				touch_hal_set_threshold(channel_cfg->channel_num,
+				(channel_cfg->channel_sens * benchmark_value / 100));
```

Comment:
the indentation is off, plus I don't think you need the outher parenthesis

Reply:
Good points